dist/tools/ethos/ethos.c

```diff
@@ -241,27 +250,40 @@ static int _serial_handle_byte(serial_t *serial, char c)
 
 static void _write_escaped(int fd, char* buf, ssize_t n)
 {
-    uint8_t out[2];
+    /*
+     * Certain USB-to-UART adapters/drivers will immediately send a USB packet
+     * with a single byte instead of buffering internally when the application
+     * does writes one byte at a time. Since USB Full Speed can only send 1
+     * packet per 1 ms, this causes huge latencies for the network, because each
+     * byte of data will then add at least 1 ms on the latency.
+     * Observed on NXP OpenSDAv2 (Kinetis FRDM boards), both CMSIS/mbed DAPlink
+     * and Segger Jlink firmware are affected.
+     */
+    /* Our workaround is to prepare the data to send in a local buffer and then
+     * call write() on the buffer instead of one char at a time */
+    uint8_t out[SERIAL_BUFFER_SIZE];
     size_t escaped = 0;
+    size_t buffered = 0;
 
     while(n--) {
-        char c = *buf++;
-        switch(c) {
-            case LINE_FRAME_DELIMITER:
-                out[0] = LINE_ESC_CHAR;
-                out[1] = (LINE_FRAME_DELIMITER ^ 0x20);
-                checked_write(fd, out, 2);
-                escaped++;
-                break;
-            case LINE_ESC_CHAR:
-                out[0] = LINE_ESC_CHAR;
-                out[1] = (LINE_ESC_CHAR ^ 0x20);
-                checked_write(fd, out, 2);
-                escaped++;
-                break;
-            default:
-                checked_write(fd, &c, 1);
+        unsigned char c = *buf++;
+        if (c == LINE_FRAME_DELIMITER || c == LINE_ESC_CHAR) {
+            out[buffered++] = LINE_ESC_CHAR;
+            c ^= 0x20;
+            ++escaped;
+            if (buffered >= SERIAL_BUFFER_SIZE) {
```

Comment:
would it make sense to factor out these three lines into a static function?
```
static void _maybe_write(fd, out, *buffered, threshold)
{
  if (*buffered >= threshold) {
     checked_write(fd, out, *buffered);
     *buffered = 0;
  }
}
```

Could be used three times.